File: adventjs.dev/2025/04.ts
function decodeSantaPin(code: string): string | null {
  const blocks = [...code.matchAll(/\[([^\]]+)\]/g)];
  if (blocks.length !== 4) {
    return null;
  }

  const values: number[] = [];
  blocks.forEach(([_, block], i) => {
    const rawNumber = block[0];
    let currentNumber =
      rawNumber === "<" ? values[i - 1] : parseInt(rawNumber, 10);

    for (let j = 1; j < block.length; ++j) {
      const offset = block[j] === "+" ? 1 : -1;
      currentNumber += offset;
    }

    values.push(currentNumber);
  });

  const mod10 = (n: number) => ((n % 10) + 10) % 10;
  return values.map(mod10).join("");
}

File: adventjs.dev/2025/04.py
def decode_santa_pin(code: str) -> str | None:
    import re

    blocks = re.findall(r"\[([^\]]+)\]", code)
    if len(blocks) != 4:
        return None

    values = []
    for i, block in enumerate(blocks):
        raw_number = block[0]
        if raw_number == "<":
            current_number = values[i - 1]
        else:
            current_number = int(raw_number)

        for operation in block[1:]:
            offset = 1 if operation == "+" else -1
            current_number += offset

        values.append(current_number)

    return "".join([v % 10 for v in values])
